boxtree/tree_build.py

```diff
@@ -46,6 +58,273 @@ class MaxLevelsExceeded(RuntimeError):
+                   box_child_ids=box_children, box_levels=box_levels)
+
+
+def _sort_boxes_by_level(tob, queue=None):
+    if np.any(np.diff(tob.box_levels) < 0):
+        # reorder boxes to into non-decreasing levels
+        neworder = np.argsort(tob.box_levels)
+        box_centers = tob.box_centers[:, neworder]
+        box_parent_ids = tob.box_parent_ids[neworder]
+        box_child_ids = tob.box_child_ids[:, neworder]
+        box_levels = tob.box_levels[neworder]
+        tob = TreeOfBoxes(
+            box_centers=box_centers, root_extent=tob.root_extent,
+            box_parent_ids=box_parent_ids, box_child_ids=box_child_ids,
+            box_levels=box_levels)
+    return tob
+
+
+def _sort_and_prune_deleted_boxes(tob):
+    tob = _sort_boxes_by_level(tob)
+    n_stale_boxes = np.sum(tob.box_levels == np.inf)
+    newn = tob.nboxes - n_stale_boxes
+    from dataclasses import replace
+    return replace(tob, box_parent_ids=tob.box_parent_ids[:newn],
+                   box_child_ids=tob.box_child_ids[:, :newn],
+                   box_levels=tob.box_levels[:newn],
+                   box_centers=tob.box_centers[:, :newn])
+
+
+def refined_and_coarsened(tob: TreeOfBoxes,
+                          refine_flags: Union[None, npt.NDArray] = None,
+                          coarsen_flags: Union[None, npt.NDArray] = None,
+                          error_on_ignored_flags: bool = True
+                          ) -> TreeOfBoxes:
+    """Make a refined/coarsened copy. When children of the same parent box
+    are marked differently, the refinement flag takes priority.
+
+    Both refinement and coarsening flags can only be set of leaves.
+    To prevent drastic mesh change, coarsening is only executed when a leaf
+    box is marked for coarsening, and its parent's children are all leaf
+    boxes (so that change in the number of boxes is bounded per box flagged).
+    Please note that the above behavior may be subject to change in the future.
+
+    :arg refine_flags: a boolean array of size `nboxes`.
+    :arg coarsen_flags: a boolean array of size `nboxes`.
+    :arg error_on_ignored_flags: if true, an exception is raised when enforcing
+        level restriction requires ignoring some coarsening flags.
+    :returns: a processed copy of the tree.
+    """
+    if refine_flags is None:
+        refine_flags = np.zeros(tob.nboxes, bool)
+    if coarsen_flags is None:
+        coarsen_flags = np.zeros(tob.nboxes, bool)
+
+    if (refine_flags & coarsen_flags).any():
+        raise ValueError("some boxes are simultaneously marked "
+                         "to refine and coarsen")
+
+    tob = _apply_refine_flags_without_sorting(refine_flags, tob)
+    coarsen_flags = _resized_array(coarsen_flags, tob.nboxes)
+    tob = _apply_coarsen_flags(coarsen_flags, tob, error_on_ignored_flags)
+    return _sort_and_prune_deleted_boxes(tob)
+
+# }}} End refine/coarsen a tree of boxes
+
+
+def make_tob_root(dim: int,
```

Comment:
The name isn't great, because it's not discoverable (e.g. via Ctrl-F `TreeOfBoxes`). `make_tree_of_root_box` maybe?